plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,329 @@
+
+Furthermore, moving a strict binding is much more likely to increase the cost of the
+program. Floating a strict binding into the RHS of another binding, or into the
+argument of an application, can cause its RHS to be evaluated multiple times.
+See Note [Where can a binding be floated into?] for examples. We therefore only move
+strict bindings whose RHS is trivial to evaluated (such as a variable or a
+lambda abstraction). Since being trivial to evaluate implies being effect-free, we
+only need to check the former.
+-}
+
+{- Note [Where can a binding be floated into?]
+
+TL;DR: Floating a binding into the RHS of another binding, the argument of an application,
+or a (term or type) lambda abstraction, can all lead to increased costs, even if
+we only float lazy bindings, and strict bindings that are trivial to evaluate.
+
+The reason for the cost increase is the same as why floating bindings inwards can
+save cost: when we are moving a bind `let x (nonstrict) = rhs`, what is really
+being moved around is `delay rhs` and lambda applications, which are not free.
+
+We choose to avoid floating bindings into the RHS of another binding. By doing so,
+floating bindings into lambda abstractions can no longer increase costs. On the other
+hand, we _do_ float bindings into arguments of applications. This may lead to increased
+costs, but when it occurs, the increase is almost always small compared to the total
+cost of the program.
+
+Next we explain the above statements in details.
+
+==== Floating a binding into the RHS of another binding ====
+
+Here's an example where floating a binding into the RHS of another binding leads to
+increased cost due to the CEK machine taking more steps to evaluate the program:
+
+let x (nonstrict) = rhs_x
+ in let y = x+x in y+y
+
+Without floating, it is compiled into
+
+[(\x -> [(\y -> y+y) (delay (x+x))] (delay rhs_x)]
+
+and evaluating this program involves two `Delay` steps, one for each `delay`.
+
+If we float `x` into the RHS of `y`, it will be compiled into
+
+[(\y -> y+y) (delay [(\x -> x+x) (delay rhs_x)])]
+
+and evaluating this program involves three `Delay` steps, one for the first `delay` and
+two for the second `delay`. Note that the second `delay` is inside the first `delay`.
+
+Again we are using pseudo-UPLC here for better readability, e.g., we omitted such things
+as `force`. The number of `Force` steps are the same with or without floating in,
+which follows from the fact that `rhs_x` is always evaluated the same number of times
+however we do the floating.
+
+==== Floating a binding into the argument of an application ====
+
+Since floating a binding into the RHS of another binding can lead to increased costs,
+it follows that floating a binding into the argument of an application can as well,
+because `let x (nonstrict) = rhs in body` is the same as `(\x -> force body) (delay rhs)`,
+and `let x (strict) = rhs in body` is the same as `(\x -> body) rhs`.
```

Comment:
Did I say that floating into arguments was bad? I did, but not in my last comment... I think it's actually _fine_, because of the strictness.

In `[ (\x. .. x .. x ..) t ]` t gets evaluated _once_, even though the resulting value is used multiple times. The problem only comes if we are in a context where we get used multiple times _and_ we've got delayed evaluation. So I continue to think (as I said here: https://github.com/input-output-hk/plutus/pull/5032#discussion_r1065703355) that what really matters is _which constructs delay evaluation_. And that's lambdas, type abstractions, and non-strict let-binding RHSs.

Which is to say: if we do the right thing there, then floating into arguments is fine.